src/types.rs

```diff
@@ -165,13 +165,14 @@ impl TryFrom<HWIDeviceInternal> for HWIDevice {
     type Error = Error;
     fn try_from(h: HWIDeviceInternal) -> Result<HWIDevice, Error> {
         match h.error {
-            Some(e) => match h.code {
-                Some(code) => match ErrorCode::try_from(code) {
-                    Ok(code) => Err(Error::HWIErrorWithCode { msg: e, code }),
-                    Err(_) => Err(Error::HWIError(e)),
-                },
-                None => Err(Error::HWIError(e)),
-            },
+            Some(e) => {
+                if let Some(code) = h.code {
+                    if let Ok(code) = ErrorCode::try_from(code) {
+                        return Err(Error::HWIError(e, Some(code)));
+                    }
+                }
+                Err(Error::HWIError(e, None))
+            }
```

Comment:
This could be re-written as 
```
                let code = h.code.map(|c| ErrorCode::try_from(c).ok()).flatten();
                Err(Error::HWIError(e, code))
```

Reply:
Ok ok, in the next commits I'll follow your suggestion. Thank you ☺️